multi_aocx tutorial initial implementation (#1)
renlizha/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/multi_aocx/device_code_split/README.md

| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel&reg; FPGA 3rd party / custom platforms with oneAPI support <br> **Note**: Intel&reg; FPGA PAC hardware is only compatible with Ubuntu 18.04*
| Software                          | Intel® oneAPI DPC++/C++ Compiler
| What you will learn               | How to use the `-fsycl-device-code-split=per_source|per_kernel` to create multiple aocx.
| Time to complete                  | 30 minutes

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

## Purpose

> `-fsycl-device-code-split=[per_kernel|per_source]` flag could be used when a design doesn't fit a single FPGA, so user must split it into multiple smaller images.

Reviewer: You can keep this section simple:
The purpose of this tutorial is to describe the -fsycl-device-code-split flag and to show how it can be leveraged to break down the compilation of several kernels across different files into smaller individual compiles.

Then start a new section, something like "Using the flag"

Start by describing what this flag does. Eg:

_-fsycl-device-code-split specifies how the device code in the input set of source files will be structured for compilation._

This should be followed by what the different options do and you can take the text from the flag documentation.
Then, follow up with what the benefit is - if all kernels do not fit then we can ask the compiler generate smaller set of images.

> ### Simple Code Example

Reviewer: This is not a code example, you can remove this section and include its contents in the 'Using the flag' section I described above.

```
icpx ... simple.cpp -o simple.exe -fsycl-device-code-split=per_kernel

> icpx ... -reuse-exe=simple.exe -Xsv

Reviewer: -reuse-exe has been used here without any description. Same as my comments above:
- Mention in the purpose section that this flag will be covered.
- Describe what it does
- Describe why it is useful
- Then show this example of how to use it and how it helps

```

> ### Using the device_code_split

Reviewer: This section should be renamed and restructured as I mentioned above.

- Split a large design into smaller images
  - `per_kernel` - a separate device code module is created for each SYCL kernel
  - `per_source` - a separate device code module is created for each source

> - Users don't need to restructure their code 

Reviewer: I think you can elaborate more on this:
_Users do not need to manually restructure their code or source files to get smaller hardware image files. This flag allows the compiler to do so automatically._

> - All aocxs need to be recompiled if any code or target changed

Reviewer: Elaborate here as well:
_When using the -dsycl-device-code-split, all kernels will be recompiled if any code or the target is changed unless we use the -reuse-exe flag to specify to the compiler which hardware images can be reused._

And make sure the reuse-exe flag is described before you mention this.

> ## Key Concepts

Reviewer: Do not keep any empty sections. You can fill this out 1 line summaries for these flags.

You can compile and run this tutorial in the Eclipse*IDE (in Linux*) and the Visual Studio*IDE (in Windows*).
For instructions, refer to [FPGA Workflows on Third-Party IDEs for Intel&reg; oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html).

## Examining the Reports

Locate the pair of `report.html` files in either:

- **Report-only compile**:  `device_code_split_report.prj`
- **FPGA hardware compile**: `device_code_split.prj`

Scroll down on the Summary page of the report and expand the section titled 

## Running the Sample

1. Run the sample on the FPGA emulator (the kernel executes on the CPU):

   ```bash
   ./device_code_split.fpga_emu    (Linux)
   device_code_split.fpga_emu.exe  (Windows)
   ```

2. Run the sample of the FPGA simulator device

   ```bash
   ./device_code_split.fpga_sim        (Linux)
   device_code_split.fpga_sim.exe      (Windows)
   ```

3. Run the sample on the FPGA device

   ```bash
   ./device_code_split.fpga             (Linux)
   device_code_split.fpga.exe           (Windows)
   ```

### Example of Output on Emulator

> ### Discussion of Results

Reviewer: You can show the difference in area utilization in the reports and show how per_kernel generated image uses lesser area than per_source if the tutorial is doing both compiles.

Reviewer: Hey Renli this was marked as resolved but is still not addressed.

| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel&reg; FPGA 3rd party / custom platforms with oneAPI support <br> **Note**: Intel&reg; FPGA PAC hardware is only compatible with Ubuntu 18.04*
| Software                          | Intel® oneAPI DPC++/C++ Compiler
| What you will learn               | How to use the `-fsycl-device-code-split=per_source|per_kernel` to create multiple aocx.
| Time to complete                  | 30 minutes

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

## Purpose

> The purpose of this tutorial is to describe the `-fsycl-device-code-split` flag and to show how it can be leveraged to break down the compilation of several kernels across different files into smaller individual compiles. 

Reviewer: ```suggestion
The purpose of this tutorial is to describe the `-fsycl-device-code-split` flag and to show how it can be leveraged to break down the compilation of several kernels into smaller individual compiles. 
```

> The tutorial will also include `-reuse-exe=<exe>` flag and shows how it can enable rapid recompilation of host-only code changes

Reviewer: ```suggestion
The tutorial will also describe the `-reuse-exe=<exe>` flag and show how it can enable rapid recompilation of host-only code changes.
```

### Using the flag

`-fsycl-device-code-split=[per_kernel|per_source]` specifies how the device code in the input set of source files will be structured for compilation.

- It splits a large design into smaller images

>   - `per_kernel` - a separate device code module is created for each SYCL kernel

Reviewer: Sentences are missing full stops in many places, please add them. This type of stuff will also be brought up in external review so might as well address it now

  - `per_source` - a separate device code module is created for each source
- Users do not need to manually restructure their code or source files to get smaller hardware image files. This flag allows the compiler to do so automatically. 

> - When using the `-fsycl-device-code-split`, all kernels will be recompiled if any code or the target is changed unless we use the `-reuse-exe` flag (described below) to specify to the compiler which hardware images can be reused.

Reviewer: ```suggestion
- When using the `-fsycl-device-code-split` flag, all kernels will be recompiled if any code or the target is changed unless we use the `-reuse-exe` flag (described below) to specify to the compiler which hardware images can be reused.
```

> `-reuse-exe=<exe>` re-uses the device binary embedded within the previously compiled SYCL binary <exe>

Reviewer: ```suggestion
`-reuse-exe=<exe>` re-uses the device binary embedded within the previously compiled SYCL binary named `<exe>`.
```

> - It speeds up FPGA aoc compile if the decie code in <exe> is unchanged

Reviewer: ```suggestion
- It speeds up the FPGA compilation step if the device code corresponding to `<exe>` is unchanged.
```

- It can minimize or avoid long Quartus compile times for FPGA targets
- This only works when the device code is unchanged

> using the flags together:

Reviewer: ```suggestion
You can use the flags together as shown below:
```

```
icpx ... simple.cpp -o simple.exe -fsycl-device-code-split=per_kernel
...
icpx ... -reuse-exe=simple.exe -Xsv
```

## Key Concepts

> - Using the `-fsycl-device-code-split=[per_kernel|per_source]` flag allows the compiler to split a larger design into smaller hardware images automatically, saving users' time from manual restructure.

Reviewer: ```suggestion
- Using the `-fsycl-device-code-split=[per_kernel|per_source]` flag instructs the compiler to split a larger design into smaller hardware images automatically, saving users' time from manually restructuring the code.
```

> - If the device code and options affecting the device have not changed since the previous compilation, passing the `-reuse-exe=<exe_name>` flag instructs the compiler to extract the compiled FPGA binary from the existing executable and package it into the new executable, saving the device compilation time.

Reviewer: ```suggestion
- If the device code and options affecting the device code have not changed since the previous compilation, passing the `-reuse-exe=<exe_name>` flag instructs the compiler to extract the compiled FPGA binary from the existing executable and package it into the new executable, saving the device code compilation time.
```

### In Third-Party Integrated Development Environments (IDEs)

You can compile and run this tutorial in the Eclipse*IDE (in Linux*) and the Visual Studio*IDE (in Windows*).
For instructions, refer to [FPGA Workflows on Third-Party IDEs for Intel&reg; oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html).

## Examining the Reports

Locate the pair of `report.html` files in either:

- **Report-only compile**:  `device_code_split_report.prj`
- **FPGA hardware compile**: `device_code_split.prj`

Scroll down on the Summary page of the report and expand the section titled 

## Running the Sample

1. Run the sample on the FPGA emulator (the kernel executes on the CPU):

   ```bash
   ./device_code_split.fpga_emu    (Linux)
   device_code_split.fpga_emu.exe  (Windows)
   ```

2. Run the sample of the FPGA simulator device

   ```bash
   ./device_code_split.fpga_sim        (Linux)
   device_code_split.fpga_sim.exe      (Windows)
   ```

3. Run the sample on the FPGA device

   ```bash
   ./device_code_split.fpga             (Linux)
   device_code_split.fpga.exe           (Windows)
   ```

> ### Example of Output on Emulator

Reviewer: Remove this section if there is nothing relevant to display. Infact we should think of removing the emulation section from the Readme and the CMake, if the flags have no effect on emulation.

| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel&reg; FPGA 3rd party / custom platforms with oneAPI support <br> **Note**: Intel&reg; FPGA PAC hardware is only compatible with Ubuntu 18.04*
| Software                          | Intel® oneAPI DPC++/C++ Compiler
| What you will learn               | How to use the `-fsycl-device-code-split=per_source|per_kernel` to create multiple aocx.
| Time to complete                  | 30 minutes

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

## Purpose

> The purpose of this tutorial is to describe the `-fsycl-device-code-split` flag and to show how it can be leveraged to break down the compilation of several kernels into smaller individual compiles.  

Reviewer 2: I think we should clarify this a bit. 

This options splits and generates multiple FPGA bitstreams. The total compile time will likely be longer, but the main perk is if your image is getting too large to fit on the FPGA, this can be a nice way to split them.

Author: I think the "Purpose" section is just a one line description of what it does, the pro and cons can be added to the "Using the flag" section?

Reviewer 2: Okay - that sounds good. Is this information already, or do we want to re-iterate it there?

The tutorial will also describe the `-reuse-exe=<exe>` flag and show how it can enable rapid recompilation of host-only code changes.

### Using the flag

`-fsycl-device-code-split=[per_kernel|per_source]` specifies how the device code in the input set of source files will be structured for compilation.

- It splits a large design into smaller images:
  - `per_kernel` - a separate device code module is created for each SYCL kernel.
  - `per_source` - a separate device code module is created for each source.
- Users do not need to manually restructure their code or source files to get smaller hardware image files. This flag allows the compiler to do so automatically. 

> - When using the `-fsycl-device-code-split` flag, all kernels will be recompiled if any code or the target is changed unless we use the `-reuse-exe` flag (described below) to specify to the compiler which hardware images can be reused.

Reviewer 2: Use third person language. So instead of "unless we use the...", say "unless you use the..."

`-reuse-exe=<exe>` re-uses the device binary embedded within the previously compiled SYCL binary <exe>.

- It speeds up the FPGA compilation step if the device code corresponding to `<exe>` is unchanged.
- It can minimize or avoid long Quartus compile times for FPGA targets.
- This only works when the device code is unchanged.

using the flags together:
```

> icpx ... simple.cpp -o simple.exe -fsycl-device-code-split=per_kernel

Reviewer 2: I don't think you need both examples. In fact, you can always include `-reuse-exe`. If it can't find a binary, it will just recompile. So I think you need just the second example. 

Btw, do you know if you need to specify the exe name? I thought it defaulted to picking it up from the `-o` flag?

Author: Yes, user must specify a exe name or else it will throw "Option -reuse-exe=<exe> requires a filename" error.

> # Using the flows test 'device_code_split'

Reviewer 3: ```suggestion
#  'device_code_split' Sample
```

> This FPGA tutorial demonstrates the -fsycl-device-code-split=per_source|per_kernel
> 
> | Optimized for                     | Description
> |:---                               |:---
> | OS                                | Linux* Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
> | Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel&reg; FPGA 3rd party / custom platforms with oneAPI support <br> **Note**: Intel&reg; FPGA PAC hardware is only compatible with Ubuntu 18.04*
> | Software                          | Intel® oneAPI DPC++/C++ Compiler
> | What you will learn               | How to use the `-fsycl-device-code-split=per_source|per_kernel` to create multiple aocx.
> | Time to complete                  | 30 minutes
> 
> > **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
> >
> > For using the simulator flow, one of the following simulators must be installed and accessible through your PATH:
> > - Questa*-Intel® FPGA Edition
> > - Questa*-Intel® FPGA Starter Edition
> > - ModelSim® SE
> >
> > When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
> 
> ## Purpose
> The purpose of this tutorial is to describe the `-fsycl-device-code-split` flag and to show how it can be leveraged to break down the compilation of several kernels into smaller individual compiles.
> 
> The tutorial will also describe the `-reuse-exe=<exe>` flag and show how it can enable rapid recompilation of host-only code changes.

Reviewer 3: ````suggestion
This sample is an FPGA tutorial that demonstrates the `-fsycl-device-code-split=per_source|per_kernel`

| Area                 | Description
|:--                      |:--
| What you will learn  | How to use the `-fsycl-device-code-split=per_source|per_kernel` compiler flag to create multiple aocx
| Time to complete     | 20 minutes
| Category             | Concepts and Functionality

## Purpose
This FPGA tutorial shows how to use the `-fsycl-device-code-split` compiler flag and how it can be leveraged to break down the compilation of several kernels into smaller individual compiles.

The tutorial will also describe the `-reuse-exe=<exe>` flag and show how it can enable rapid recompilation of host-only code changes.

## Prerequisites

| Optimized for       | Description
|:---                         |:---
| OS                        | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
| Hardware            | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software            | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

> **Warning** Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

This sample is part of the FPGA code samples.
It is categorized as a Tier 2 sample that demonstrates a compiler feature.

```mermaid
flowchart LR
   tier1("Tier 1: Get Started")
   tier2("Tier 2: Explore the Fundamentals")
   tier3("Tier 3: Explore the Advanced Techniques")
   tier4("Tier 4: Explore the Reference Designs")

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```
````

> ### Using the flag

Reviewer 3: ```suggestion
### Details on how to use the `-fsycl-device-code-split` flag
```

> `-fsycl-device-code-split=[per_kernel|per_source]` specifies how the device code in the input set of source files will be structured for compilation.

Reviewer 3: ```suggestion
The `-fsycl-device-code-split=[per_kernel|per_source]` compiler flag specifies how the device code in the input set of source files will be structured for compilation.
```

> - It splits a large design into smaller images:

Reviewer 3: ```suggestion
- It makes the compiler split a large design into smaller images:
```

  - `per_kernel` - a separate device code module is created for each SYCL kernel.
  - `per_source` - a separate device code module is created for each source.
- Users do not need to manually restructure their code or source files to get smaller hardware image files. This flag allows the compiler to do so automatically. 
- When using the `-fsycl-device-code-split` flag, all kernels will be recompiled if any code or the target is changed unless you use the `-reuse-exe` flag (described below) to specify to the compiler which hardware images can be reused.
- The total compile time will likely be longer, but the main perk is if your image is getting too large to fit on the FPGA, this can be a nice way to split them.

> `-reuse-exe=<exe>` re-uses the device binary embedded within the previously compiled SYCL binary <exe>.

Reviewer 3: ```suggestion
The `-reuse-exe=<exe>` compiler flag makes the compiler re-use the device binary embedded within the previously compiled SYCL binary <exe>.
```

- It speeds up the FPGA compilation step if the device code corresponding to `<exe>` is unchanged.
- It can minimize or avoid long Quartus compile times for FPGA targets.
- This only works when the device code is unchanged.

> 

Reviewer 3: ```suggestion
The [fast_recompile](/DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fast_recompile) sample covers this topic in details.

```

using the flags together:
```
icpx ... simple.cpp -o simple.exe -fsycl-device-code-split=per_kernel -reuse-exe=simple.exe
```
If `simple.exe` does not exist, `-reuse-exe` is ignored and the FPGA device image is regenerated. This will always be the case the first time a project is compiled.

## Key Concepts
- Using the `-fsycl-device-code-split=[per_kernel|per_source]` flag allows the compiler to split a larger design into smaller hardware images automatically, saving users' time from manual restructure.
- If the device code and options affecting the device have not changed since the previous compilation, passing the `-reuse-exe=<exe_name>` flag instructs the compiler to extract the compiled FPGA binary from the existing executable and package it into the new executable, saving the device compilation time.

> ## Building the `device_code_split` Tutorial
> 
> > **Note**: If you have not already done so, set up your CLI
> > environment by sourcing  the `setvars` script located in
> > the root of your oneAPI installation.
> >
> > Linux*:
> >
> > - For system wide installations: `. /opt/intel/oneapi/setvars.sh`
> > - For private installations: `. ~/intel/oneapi/setvars.sh`
> >
> > Windows*:
> >
> > - `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
> >
> >For more information on environment variables, see **Use the setvars Script** for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
> 
> ### Running Samples in Intel&reg; DevCloud
> 
> If you are running a sample in the Intel&reg; DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode:
> 
> - Compiles to FPGA are supported only on `fpga_compile` nodes.
> - Executing programs on FPGA hardware is supported only on `fpga_runtime` nodes of the appropriate type, such as `fpga_runtime:arria10` or `fpga_runtime:stratix10`.
> 
> On the login nodes, you cannot compile or execute programs on FPGA hardware. For more information, see the Intel&reg; oneAPI Base Toolkit Get Started Guide ([https://devcloud.intel.com/oneapi/documentation/base-toolkit/](https://devcloud.intel.com/oneapi/documentation/base-toolkit/)).
> 
> When compiling for FPGA hardware, increase the job timeout to 12h.
> 
> ### Using Visual Studio Code*  (Optional)
> 
> You can use Visual Studio Code (VS Code) extensions to set your environment, create launch configurations,
> and browse and download samples.
> 
> The basic steps to build and run a sample using VS Code include:
> 
> - Download a sample using the extension **Code Sample Browser for Intel&reg; oneAPI Toolkits**.
> - Configure the oneAPI environment with the extension **Environment Configurator for Intel&reg; oneAPI Toolkits**.
> - Open a Terminal in VS Code (**Terminal>New Terminal**).
> - Run the sample in the VS Code terminal using the instructions below.
> 
> To learn more about the extensions and how to configure the oneAPI environment, see the
> [Using Visual Studio Code with Intel&reg; oneAPI Toolkits User Guide](https://software.intel.com/content/www/us/en/develop/documentation/using-vs-code-with-intel-oneapi/top.html).

Reviewer 3: ```suggestion
## Build the `device_code_split` Tutorial

>**Note**: When working with the command-line interface (CLI), you should configure the oneAPI toolkits using environment variables. Set up your CLI environment by sourcing the `setvars` script in the root of your oneAPI installation every time you open a new terminal window. This practice ensures that your compiler, libraries, and tools are ready for development.
>
> Linux*:
> - For system wide installations: `. /opt/intel/oneapi/setvars.sh`
> - For private installations: ` . ~/intel/oneapi/setvars.sh`
> - For non-POSIX shells, like csh, use the following command: `bash -c 'source <install-dir>/setvars.sh ; exec csh'`
>
> Windows*:
> - `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
> - Windows PowerShell*, use the following command: `cmd.exe "/K" '"C:\Program Files (x86)\Intel\oneAPI\setvars.bat" && powershell'`
>
> For more information on configuring environment variables, see [Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html) or [Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
```

> ### On a Linux* System

Reviewer 3: ```suggestion
### On Linux*
```

> 1. Install the design in `build` directory from the design directory by running `cmake`:
> 
>    ```bash
>    mkdir build
>    cd build
>    ```
> 
>    If you are compiling for the Intel&reg; PAC with Intel Arria&reg; 10 GX FPGA, run `cmake` using the command:
> 
>    ```bash
>    cmake ..
>    ```
> 
>    Alternatively, to compile for the Intel&reg; FPGA PAC D5005 (with Intel Stratix&reg; 10 SX), run `cmake` using the command:
> 
>    ```bash
>    cmake .. -DFPGA_DEVICE=intel_s10sx_pac:pac_s10
>    ```
> 
>    You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
> 
>    ```bash
>    cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
>    ```
> 
> 2. Compile the design using the generated `Makefile`. The following four build targets are provided that match the recommended development flow:
> 
>    - Compile and run for emulation (fast compile time, targets emulates an FPGA device) using:
> 
>      ```bash
>      make fpga_emu
>      ```
> 
>    - Generate HTML optimization reports using:
> 
>      ```bash
>      make report
>      ```
> 
>    - Compile for simulation (fast compile time, targets simulated FPGA device)
> 
>      ```bash
>      make fpga_sim
>      ```
> 
>    - Compile and run on FPGA hardware (longer compile time, targets an FPGA device) using:
> 
>      ```bash
>      make fpga
>      ```
> 
> 3. (Optional) As the earlier hardware compile can take several hours to complete, FPGA precompiled binaries (compatible with Ubuntu 18.04) can be downloaded [here](https://iotdk.intel.com/fpga-precompiled-binaries/latest/device_code_split.fpga.tar.gz).

Reviewer 3: ````suggestion
1. Change to the sample directory.
2. Build the program for Intel® Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake ..
   ```
   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```
   >
   > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
   >  ```
   >
   > You will only be able to run an executable on the FPGA if you specified a BSP.

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile and run for emulation (fast compile time, targets emulates an FPGA device).
      ```
      make fpga_emu
      ```
   2. Generate the HTML optimization reports. (See [Read the Reports](#read-the-reports) below for information on finding and understanding the reports.)
      ```
      make report
      ```
   3. Compile for simulation (fast compile time, targets simulated FPGA device).
      ```
      make fpga_sim
      ```
   4. Compile and run on FPGA hardware (longer compile time, targets an FPGA device).
      ```
      make fpga
      ```
````

> ### On a Windows* System
> 
> 1. Generate the `Makefile` by running `cmake`.
> 
>    ```
>    mkdir build
>    cd build
>    ```
> 
>    To compile for the Intel&reg; PAC with Intel Arria&reg; 10 GX FPGA, run `cmake` using the command:
> 
>     ```
>     cmake -G "NMake Makefiles" ..
>    ```
> 
>    Alternatively, to compile for the Intel&reg; FPGA PAC D5005 (with Intel Stratix&reg; 10 SX), run `cmake` using the command:
> 
>    ```
>    cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=intel_s10sx_pac:pac_s10
>    ```
> 
>    You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
> 
>    ```
>    cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
>    ```
> 
> 2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:
> 
>    - Compile for emulation (fast compile time, targets emulated FPGA device):
> 
>      ```
>      nmake fpga_emu
>      ```
> 
>    - Generate the optimization report:
> 
>      ```
>      nmake report
>      ```
> 
>    - Compile for simulation (fast compile time, targets simulated FPGA device, reduced problem size):
> 
>      ```
>      nmake fpga_sim
>      ```
> 
>    - Compile for FPGA hardware (longer compile time, targets FPGA device):
> 
>      ```
>      nmake fpga
>      ```
> 
> > **Note**: The Intel&reg; PAC with Intel Arria&reg; 10 GX FPGA and Intel&reg; FPGA PAC D5005 (with Intel Stratix&reg; 10 SX) do not yet support Windows*. Compiling to FPGA hardware on Windows* requires a third-party or custom Board Support Package (BSP) with Windows* support.
> 
> > **Note**: If you encounter any issues with long paths when compiling under Windows*, you might have to create your `build` directory in a shorter path, for example `c:\samples\build`.  You can then run `cmake` from that directory, and provide `cmake` with the full path to your sample directory.

Reviewer 3: ````suggestion
### On Windows*

1. Change to the sample directory.
2. Build the program for the Intel® Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake -G "NMake Makefiles" ..
   ```
   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```
   >
   > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command:
   >  ```
   >  cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
   >  ```
   >
   > You will only be able to run an executable on the FPGA if you specified a BSP.

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile for emulation (fast compile time, targets emulated FPGA device).
      ```
      nmake fpga_emu
      ```
   2. Generate the optimization report. (See [Read the Reports](#read-the-reports) below for information on finding and understanding the reports.)
      ```
      nmake report
      ```
   3. Compile for simulation (fast compile time, targets simulated FPGA device, reduced problem size).
      ```
      nmake fpga_sim
      ```
   4. Compile for FPGA hardware (longer compile time, targets FPGA device):
      ```
      nmake fpga
      ```
> **Note**: If you encounter any issues with long paths when compiling under Windows*, you may have to create your ‘build’ directory in a shorter path, for example c:\samples\build.  You can then run cmake from that directory, and provide cmake with the full path to your sample directory.
````

> ### In Third-Party Integrated Development Environments (IDEs)
> 
> You can compile and run this tutorial in the Eclipse*IDE (in Linux*) and the Visual Studio*IDE (in Windows*).
> For instructions, refer to [FPGA Workflows on Third-Party IDEs for Intel&reg; oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html).

Reviewer 3: ```suggestion
```

> ## Examining the Reports

Reviewer 3: ```suggestion
## Read the Reports
```

Locate the pair of `report.html` files in either:

- **Report-only compile**:  `device_code_split_report.prj`
- **FPGA hardware compile**: `device_code_split.prj`

> Scroll down on the Summary page of the report and expand the section titled 

Reviewer 3: The sentence is not finished here

> ## Running the Sample
> 
> 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
> 
>    ```bash
>    ./device_code_split.fpga_emu    (Linux)
>    device_code_split.fpga_emu.exe  (Windows)
>    ```
> 
> 2. Run the sample of the FPGA simulator device
> 
>    ```bash
>    CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./device_code_split.fpga_sim    (Linux)
>    CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./device_code_split.fpga_sim.exe  (Windows)
>    ```
> 
> 3. Run the sample on the FPGA device
> 
>    ```bash
>    ./device_code_split.fpga             (Linux)
>    device_code_split.fpga.exe           (Windows)
>    ```

Reviewer 3: ````suggestion
## Run the `device_code_split` Sample

### On Linux

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./device_code_split.fpga_emu
   ```
2. Run the sample of the FPGA simulator device (the kernel executes on the CPU).
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./device_code_split.fpga_sim
   ```
3. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   ./device_code_split.fpga
   ```

### On Windows

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   device_code_split.fpga_emu.exe
   ```
2. Run the sample of the FPGA simulator device (the kernel executes on the CPU).
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   device_code_split.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```
3. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   device_code_split.fpga.exe
   ```

````

### Example of Output on Emulator
```
RESULT: The results are correct!
```

> ### Discussion of Results

Reviewer 3: ```suggestion
### Understand the Results
```
